database/tbcd/level/upgrade.go

```
@@ -546,3 +551,66 @@ func (l *ldb) v3(ctx context.Context) error {
 
 	return nil
 }
+
+// v4 upgrade the database from v3 to v4.
+// Changes:
+// Add heighthash (L2 height - Keystone Hash) to keystone database.
+// XXX antonio add test
+func (l *ldb) v4(ctx context.Context) error {
+	log.Tracef("v4")
+	defer log.Tracef("v4 exit")
+
+	log.Infof("Upgrading database from v3 to v4")
+
+	// Index all keystones to M[height][hash] format.
+	bhs := l.pool[level.BlockHeadersDB]
+	ksdb := l.pool[level.KeystonesDB]
+	i := ksdb.NewIterator(&util.Range{Start: nil, Limit: nil}, nil)
+	defer func() { i.Release() }()
+	var records int
+	for records = 0; i.Next(); records++ {
+		key := i.Key()
+		value := i.Value()
+		if len(value) != keystoneSize {
+			if len(value) == keystoneSizeV3 {
+				// Keystone without height encoded.
+				var u [keystoneSize]byte
+				copy(u[4:], value[:])
+				value = u[:]
+			} else {
+				// Not a keystone.
+				continue
```

isn't this the _keystones database_? are we ignoring an error here? shouldn't everything be a keystone here? or do we store different types in the same db?

if so, so we prefix those types at all? this silent continue feels like a footgun

---

A few things:
1. This upgrade changes the size of keystones, so we check if their size is different
2. The new height hash can have an index but not value
3. I believe some entries always exist with no value but @marcopeereboom may have to confirm

With that said, I think adding a log here is probably safe though could be noisy.

---

ok, this feels a bit odd to me, but I get it. let's not sweat it